### PR TITLE
WT-8476 some refactorings, now that FLCS is less of a special case

### PR DESCRIPTION
**Pull request by sauclovian-wt**
The reconciliation changes are by far the largest, so I made that a separate commit.

Note that because this moves the history cursor involved to the WT_RECONCILE structure, it (potentially) reduces the number of history cursors created, because once one is created it stays around until __rec_destroy. Conversely though this means that __rec_destroy and __rec_destroy_session can now fail (though the failure case is likely theoretical)... which unfortunately reverts some of the cleanup in WT-8074 that was based on noticing that __rec_destroy_session then didn't fail. This feels untidy, but I guess it just reflects poor timing of this change vs. that one.

Also, if possible someone please double-check that I didn't drop or mess up any of the WT-8074 changes to the code I moved around, in the course of merging with it; e.g. the failpoint failure rate changed and I think I got that, but it's very easy to slip up when doing such merges.

### CODE REVIEW

**Comment by lukech** [comment]
autobuild: Can one of the admins verify this patch?

**Comment by keithbostic** [comment]
@luke-pearson, can you make sure we haven't broken WT-8074, please?

**Comment by luke-pearson** [comment]
> luke-pearson, can you make sure we haven't broken WT-8074, please?

It looks like it does indeed break WT-8074, I'll try and suggest fixes.

File: src/reconcile/rec_write.c
Comment (luke-pearson) [inline src/reconcile/rec_write.c:266]: WT-8074 should've gone in a slightly different direction to support further reconciliation changes. We should have set a boolean "panic" and had an `err:` label to catch all failures beyond the panic stage. Currently this change will break the logic added in WT-8074. If you would like I can try and implement my idea on this PR and push it.

The other options is that we fix these cases. At this stage if ret == 0 then we're not panicking. So we could do:
```
if (ret == 0 && (ret = __rec_cleanup(session, r)) != 0) {
    goto panic;
} else 
     WT_TRET(__rec_cleanup(session, r));
```

File: src/reconcile/rec_write.c
Comment (luke-pearson) [inline src/reconcile/rec_write.c:287]: This would also require fixing:
```
if (ret == 0 && (ret = __rec_destroy_session(session)  != 0) {
   goto panic;
} else
     WT_TRET(session->block_manager_cleanup(session));
```

File: src/reconcile/rec_write.c
Comment (sauclovian-wt) [reply src/reconcile/rec_write.c:266]: Hrm, I seem to have misread the code. I guess I believed the comment that says "any failure past this point is a panic" :-) ... not reading carefully enough is all too common when merging.

I suspect that the tidiest way to arrange it in the long term is to move all the code past where err gets set, or at least all of it that needs to be run even if the page reconciliation fails, to its own function (which can then do normal error handling) and then panic if that fails. But maybe not, I'm falling asleep and going to need to come back tomorrow.

Will mark this PR as a draft for the moment.

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c:266]: oh and feel free to push here if you want; AFAIK it should let you...

Comment (luke-pearson) [reply src/reconcile/rec_write.c:266]: I created a branch in your fork with a suggested change which I was going to PR in your fork. But I can't push the branch to your repro can you give me permissions?

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c:266]: I have tried to, with luck it might even have worked...

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c:266]: But I think you can push to this branch because you have access to the original repo, and it might be easier to just do that.

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c]: I pushed two more commits to another branch here: https://github.com/sauclovian-wt/wiredtiger/tree/wt-8476-morebits
The first (I think) fixes my changes to not be overtly broken; the second demonstrates the change I suggested above. Didn't want to put either here directly and maybe interfere with what you might be doing, but can easily pull either or both in.

Comment (luke-pearson) [reply src/reconcile/rec_write.c]: Yeah let's go with that refactor change its much cleaner. 👍

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c]: However, there's another point. Since I was wrong about the panic paths, this change _does_ add a new trigger for the panic, and, in fact, two of them. Do we care? I feel like there probably aren't any particularly worrisome ways for cursor_reset or cursor_close to fail, but I haven't actually gone and checked. Could eliminate some of it by closing the cursor after each page instead of trying to save it...

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c]: (Your changes are also fine with me)

Comment (luke-pearson) [reply src/reconcile/rec_write.c]: Saving the cursor does sound better than closing it after each page, I don't think that `cursor_close` or `cursor_reset` would have likely failure scenarios and can probably be safely included in our panic reasons.

Comment (luke-pearson) [reply src/reconcile/rec_write.c]: > (Your changes are also fine with me)

The refactor is definitely better though.

Comment (sauclovian-wt) [reply src/reconcile/rec_write.c]: I've pushed an updated version of it here, plus the obvious next set of tidyups. I think this means we should be more or less good to go (if nothing blows up in testing, and if nobody else spots anything in the rest of the changes), so I'm going to make it not a draft again.

Other thing I guess is how to arrange things so that the ultimate commit message after merging is coherent. I don't know how that works on the merge side (having not done that for a couple years, and then in an organization that didn't approve of merge-squash) so someone tell me what to do :-)

**Comment by sauclovian-wt** [comment]
I am slightly disturbed because I did multiple clean builds with both clang and gcc this morning and none of them tripped on the unused variable. Investigating suggests it's something wrong with cmake...

File: src/reconcile/rec_write.c
Comment (luke-pearson) [reply src/reconcile/rec_write.c]: I can merge if required, I will work with you on a reasonable commit message.

**Comment by keithbostic** [comment]
gcc this time:
```
[2021/12/01 19:43:33.300]   CC       src/btree/bt_page.lo
[2021/12/01 19:43:33.300]   CC       src/btree/bt_random.lo
[2021/12/01 19:43:33.300] In file included from ../src/include/wt_internal.h:466:0,
[2021/12/01 19:43:33.300]                  from ../src/btree/bt_cursor.c:9:
[2021/12/01 19:43:33.300] ../src/btree/bt_cursor.c: In function '__wt_cursor_valid':
[2021/12/01 19:43:33.300] ../src/include/txn_inline.h:880:5: error: 'upd' may be used uninitialized in this function [-Werror=maybe-uninitialized]
[2021/12/01 19:43:33.300]      for (; upd != NULL; upd = upd->next) {
[2021/12/01 19:43:33.300]      ^~~
[2021/12/01 19:43:33.300] ../src/btree/bt_cursor.c:181:16: note: 'upd' was declared here
[2021/12/01 19:43:33.300]      WT_UPDATE *upd;
[2021/12/01 19:43:33.300]                 ^~~
[2021/12/01 19:43:33.718]   CC       src/btree/bt_read.lo```

**Review by keithbostic — CHANGES_REQUESTED** [review]

File: src/include/reconcile.h
Comment (sauclovian-wt) [reply src/include/reconcile.h:307]: It has longer lifetime than the call to the new function. The reconcile structure isn't a great place, granted, but it does have the correct lifespan...

I don't think we don't want to go back to pasting the logic for the hs-clear flag three times. It could be moved entirely inside the new function, but at the cost of recomputing it every time (probably not important) and not being able to use it to skip the call (less clear).

The history cursor is supposed to persist from one row to the next once created (that's how the pasted-three-times logic worked) -- it could be moved inside the new function at the cost of creating and destroying it for every affected row. That seems undesirable, but maybe it's not that expensive?

Comment (keithbostic) [reply src/include/reconcile.h:307]: I can live with it.

**Review by keithbostic — APPROVED** [review]

**Comment by luke-pearson** [comment]
I'll have to do a couple rounds of patch building which will take a few hours so stay tuned.

**Review by luke-pearson — APPROVED** [review]
LGTM

**Comment by sauclovian-wt** [comment]
Thanks! Sorry about the mess.

**Pull request merged**
